### ISSUE TITLE
Configuring production environment variables

### ISSUE THREAD

**Issue by karthikiyengar** [saguijs/sagui#283]
I was wondering what would be the best way to variables like the API URL on build. Since the web server serving static files from the generated dist folder cannot process environment variables, I would be unable to do something like this.

`const url = process.env.NODE_ENV === 'production' ? 'http://production' : 'http://dev'`

I believe a reasonable way to do this would be to define a webpack plugin like so: 
```
new webpack.DefinePlugin({
    __API__: apiUrl()
})
```

I am able to inject this plugin with the sagui.config.js. Although I am facing two issues here.
1. process.env.NODE_ENV seems to be undefined when sagui build is executed. Could this be related to cross_env in some way? I am on a Mac.
2. This breaks the dev server as it is unable to parse the __API__ string that webpack replaces.

Pointers would be appreciated. Thanks!

**Comment by pirelenito** [comment]
Can you paste your `sagui.config.js` ?

**Comment by karthikiyengar** [comment]
Here's the entire config:
```
/**
 * Sagui configuration object
 * see: http://sagui.js.org/
 */
var path = require('path');
var nested = require('postcss-nested');
var postCSSModulesValues = require('postcss-modules-values');
var autoprefixer = require('autoprefixer');
var webpack = require('webpack');

module.exports = {
  pages: ['index'],
  webpack: {
    plugins: [new webpack.DefinePlugin({
      __API__: process.env.NODE_ENV === 'production' ? 'http://production.url' : 'http://localhost:3000/graphql'
    })],
    module: {
      loaders: [{
        test: /\.eot.*$/,
        loader: 'url'
      }, {
        test: /\.ttf.*$/,
        loader: 'url'
      }, {
        test: /\.woff.*$/,
        loader: 'url'
      }, {
        test: /\.svg.*$/,
        loader: 'url'
      }]
    },
    output: {
      publicPath: '/'
    },
    resolve: {
      extensions: ['.js', '.json', '.jsx', '', '.css'],
      modulesDirectories: ['node_modules'],
      root: [
        path.resolve('./src')
      ]
    },
    postcss: [
      nested,
      postCSSModulesValues,
      autoprefixer({
        browsers: [
          '>1%',
          'last 4 versions',
          'Firefox ESR',
          'not ie < 9' // React doesn't support IE8 anyway
        ]
      })
    ]
  }
};

```

Using __API__ like so in my code:
`const networkInterface = createNetworkInterface({ uri: __API__ }); // eslint-disable-line`

Here's the transpiled output for that on running npm run build (runs successfully):
`var networkInterface = (0, _apolloClient.createNetworkInterface)({ uri: (http://localhost:3000/graphql) }); // eslint-disable-line
`

I did a console.log to verify the environment variables during the build, and NODE_ENV is set to undefined. This would explain the wrong URL on transpilation.

This is the error thrown on the console if I run npm run develop after adding the plugin. It's unable to parse the __API__ variable that webpack should replace.
`Uncaught SyntaxError: Unexpected token :` 

Thanks for your interest 👍

**Comment by pirelenito** [comment]
Hi @karthikiyengar, sorry for the long time to get you a reply.

A couple of things with the configuration.

## DefinePlugin

This plugin works by simply doing a find and replace in the code. So, what you really want to do is replace `__API__` with `"http://production.url"` (notice the quotes, a String literal).

So, to correctly setup the plugin you need to to add the extra `"` around the text:

```js
plugins: [new webpack.DefinePlugin({
  __API__: process.env.NODE_ENV === 'production' ? '"http://production.url"' : '"http://localhost:3000/graphql"'
})],
```

This should fix the issue.

## Undefined NODE_ENV

This is a strange one, could you confirm that the scripts in the package.json are setting the environment variable correctly (paste them here)?

## Cleanup in the config

It is pretty cool to see how people are using Sagui! However, when we design it, we envisioned the `webpack` configuration support as an *escape hatch*, that should be use as minimum as possible. 

So here are a couple of things you could improve in the config:

* We already support loading fonts (as can see by https://github.com/saguijs/sagui/blob/v7/src/webpack/loaders/font.js), so you could probably remove the `webpack.module` config;
* The `resolve` section should also not be needed (we do everything except setting the `.css` extension), which you can work around by simply adding the extension on importing any CSS file.

Let me know what you think and if you have more questions!

**Comment by karthikiyengar** [comment]
Hi @pirelenito,

Appreciate your detailed response, especially your insights on cleaning up my config. 

**DefinePlugin:** Thank you for pointing out the missing `"` marks, it was a oversight at my end. The dev server works fine.

**NODE_ENV being undefined:** I was under the impression that the is invoked with NODE_ENV set to production (had it confused with the dist command). Again, apologies for taking up your time with this.

About the config cleanup: I'm having to use font loader overrides because I need to import some icon-fonts using @font-face url imports - which don't work with file-loader out of the box. More on this here: https://github.com/webpack/css-loader/issues/38

I did not have a lot of time to go over the nuances of the two loaders, but the url-loader did fix it for me.

Pretty cool that sagui has all I need in the resolve section, I've eliminated that from my config. All in all, Sagui has been a joy to use and has been pretty much exactly what I needed! 

Thanks for all the great work and the supportive community around it.
